### ISSUE TITLE
ARM64 Support

### ISSUE THREAD

**Issue by microsby0** [docToolchain/docker-image#26]
Need support for ARM64. Current error:
```
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
```

**Comment by rdmueller** [comment]
hm. I guess that is a problem with the java binaries.

What did you do so that I can reproduce the problem?

**Comment by microsby0** [comment]
Run the top example script from the README on an M1 mac. 

```#!/usr/bin/env bash
docker run --rm -it --entrypoint /bin/bash -v ${PWD}:/project rdmueller/doctoolchain:rc-1.4.0 \
-c "doctoolchain . $1 $2 $3 $4 $5 $6 $7 $8 $9 -PinputPath=src/main/asciidoc -PmainConfigFile=config/docToolchain.groovy && exit"
```
Its a brand new machine for me so it actually took me some time to understand exactly where the issue was. M1 setup is a whole new world 😆

**Comment by rdmueller** [comment]
argh. I guess the binary image is not an arm one. But docToolchain runs on M1 machines, I once tested it.

You might want to check out this tutorial: http://doctoolchain.org/docToolchain/v2.0.x/020_tutorial/010_Install.html

It might be that you have to force it to locally install docToolchain by using once the "local" option as first parameter like this:

    cd <your project>
    curl -Lo dtcw doctoolchain.github.io/dtcw
    chmod +x dtcw
    ./dtcw local tasks
    
this will install it locally to you home folder (you can just delete the .doctoolchain folder if you want to uninstall it).
Make sure that you have java installed before you run this (version between 8 and 14).

regarding the docker image for M1 arm architecture - someone with an M1 machine has to figure out how to do it ;-)

**Comment by microsby0** [comment]
Makes sense, I'll give that work around a try and leave this issue up and hope someone with more knowledge than me comes along.

**Comment by Morl99** [comment]
I would also like this.

In order to build multi-architecture docker images, the most important thing is, that the base image also supports the relevant platforms. This is an issue with the current openjdk 13 alpine based image. I don't know of any official alpine based java image that supports arm atm. If you are willing to migrate to the debian variant of the image and update to Java 17, you could use [this](https://hub.docker.com/layers/openjdk/library/openjdk/17.0.2-jdk/images/sha256-9d16384620e3ba9d901831e5be5f23e8fc9bc1465a819516fc7c9a68fc438c91?context=explore) image. 

I would be happy to set this up with you, once you upgraded the base image to a multi-platform image, we could leave this issue open until then and create a separate issue for upgrading to Java 17 and moving to debian?

**Comment by rdmueller** [comment]
Great. Thanx for your support.
Debian shouldn't be a big issue. 
But jdk17 is one. But it is also already next on the roadmap

**Comment by rdmueller** [comment]
@Morl99 I guess I would need an M1 chip to build such an image, wouldn't I?

anyway, can't we use a debian base image and install JDK11 or JDK14 through sdkman?

**Comment by Morl99** [comment]
Or you use GitHub Actions to build all of your images?  (:

And sure, you could craft a java image yourself, but why would you do that? Is there a good reason to stick with a no longer supported java release?

**Comment by rdmueller** [comment]
Isn't jdk11 still a supported lts?

Yeah, GitHub actions would be great for the build. Do you have a bit time?

Why stick to the old version? Because in this case, an upgrade is not too easy. Too many dependencies and breaking changes

**Comment by Morl99** [comment]
11 is, but 14 (that you use here) is not.

Sure, we can work on this together, cannot be that hard to figure out ;)

**Comment by Morl99** [comment]
Most of what we would need can be found here: [Build and push docker image GitHub action](https://github.com/marketplace/actions/build-and-push-docker-images)

We could spice it up with a semantic release that automatically calculates the next version based on conventional commits, if you are up to it. Basically this turns every commit on master into an automated release. There are a couple of GitHub Actions out there that achieve this, we would need to look into which one is best suited for this project.

**Comment by m-gora** [comment]
I might also be able to help out here. The current image is super slow on Apple Silicon.

**Comment by nicolaimainiero** [comment]
You can find a working example using github workflow in my repository : https://github.com/nicolaimainiero/docker-image feel freee to copy what you need.

**Comment by rdmueller** [comment]
does anybody know if there is somewhere a virtual M1 available so that I can remotely test things like that?

**Comment by nicolaimainiero** [comment]
I think a Raspberry Pi 4 should be sufficient to test the container on the ARM64 architecture. Possibly AWS Gravitron (https://aws.amazon.com/de/ec2/graviton/) is also an option.

**Comment by m-gora** [comment]
I tried to get something running, unfortunately it is a bit complicated and I stumbled upon countless problems.

There's a Docker blogpost how to create multiarch images: https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/

I followed this blogpost by setting up docker buildx on my mac. 
However the alpine based openjdk base image does not support arm64, so i tried switching to temurin instead. 
Then I discovered that temurin does not use alpine as base image but some debian variant, so i had to switch out all the apk related commands. 

I opted for a quick solution here, but this was just not it. A lot more research is needed here (or a multi-layer build process).

**Comment by rdmueller** [comment]
I would like to follow up on this issue.
We now already have a github action with buildx setup. And it deploys to docker-hub.
I think we are not so far away from a solution yet...
